src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs

```diff
@@ -320,40 +326,58 @@ private static void GenerateProxyTypeFromProperties(Type sourceType, Type target
                 var il = methodBuilder.GetILGenerator();
                 if (sourceProperty == null)
                 {
+                    // Return a default(T) value.
                     il.DeclareLocal(targetProperty.PropertyType);
 
-                    // Return a default(T) value.
                     il.Emit(OpCodes.Ldloca_S, 0);
                     il.Emit(OpCodes.Initobj, targetProperty.PropertyType);
 
                     il.Emit(OpCodes.Ldloc_S, 0);
                     il.Emit(OpCodes.Ret);
+                    continue;
                 }
-                else
-                {
-                    // Push 'this' and get the underlying instance.
-                    il.Emit(OpCodes.Ldarg_0);
-                    il.Emit(OpCodes.Ldfld,
-                        typeBuilder.BaseType.GetField(
-                            "Instance",
-                            BindingFlags.Instance | BindingFlags.Public));
 
-                    // Call the source property.
-                    il.EmitCall(OpCodes.Callvirt, sourceProperty.GetMethod, null);
+                il.DeclareLocal(targetProperty.PropertyType);
+                il.DeclareLocal(sourceProperty.PropertyType);
+
+                // Init variables with default(T)
+                il.Emit(OpCodes.Ldloca_S, 0);
+                il.Emit(OpCodes.Initobj, targetProperty.PropertyType);
 
-                    var @return = il.DefineLabel();
+                il.Emit(OpCodes.Ldloca_S, 1);
+                il.Emit(OpCodes.Initobj, sourceProperty.PropertyType);
 
+                // Push 'this' and get the underlying instance.
+                il.Emit(OpCodes.Ldarg_0);
+                il.Emit(OpCodes.Ldfld,
+                    typeBuilder.BaseType.GetField(
+                        "Instance",
+                        BindingFlags.Instance | BindingFlags.Public));
+
+                // Call the source property.
+                il.EmitCall(OpCodes.Callvirt, sourceProperty.GetMethod, null);
+                il.Emit(OpCodes.Stloc_S, 1);
+
+                var @return = il.DefineLabel();
+
+                // BrFalse can't handle all value types (Decimal for example), so we skip it for value types
+                // The proxy will do nothing so it's fine
```

Comment:
`The proxy will do nothing so it's fine` - actually the issue here is that we really are using `brfalse` as a null-check for reference types. If it's a value type we don't need to check for null :+1: